src/v/cluster/data_migration_backend.cc

```diff
@@ -40,27 +54,753 @@ backend::backend(
+
+    // forget about the migration if it went forward or is gone
+    auto old_it = std::as_const(_mrstates).find(id);
+    if (old_it != _mrstates.cend()) {
+        const migration_reconciliation_state& old_mrstate = old_it->second;
+        vlog(
+          dm_log.debug,
+          "migration {} old sought state is {}",
+          id,
+          old_mrstate.sought_state);
+        if (!new_maybe_metadata || new_state >= old_mrstate.sought_state) {
+            vlog(
+              dm_log.debug, "dropping migration {} reconciliation state", id);
+            drop_mrstate(old_it);
+        }
+    }
+    // create new state if needed
+    if (new_maybe_metadata) {
+        const auto& new_metadata = new_maybe_metadata->get();
+        auto sought_state = new_metadata.next_ntp_state();
+        if (sought_state.has_value()) {
+            vlog(
+              dm_log.debug, "creating migration {} reconciliation state", id);
+            auto new_it = _mrstates.emplace_hint(
+              old_it, id, sought_state.value());
+            co_await reconcile_migration(new_it->second, new_metadata);
+            need_wakeup = true;
+        }
+    }
+    // delete old advance requests
+    if (auto it = _advance_requests.find(id); it != _advance_requests.end()) {
+        if (!new_state || it->second.sought_state <= new_state) {
+            _advance_requests.erase(it);
+        }
+    }
+
+    if (_is_coordinator && need_wakeup) {
+        wakeup();
+    }
+}
+
+ss::future<> backend::process_delta(cluster::topic_table_delta&& delta) {
+    vlog(dm_log.debug, "processing topic table delta={}", delta);
+    model::topic_namespace nt{delta.ntp.ns, delta.ntp.tp.topic};
+    auto it = _topic_migration_map.find(nt);
+    if (it == _topic_migration_map.end()) {
+        co_return;
+    }
+    auto migration_id = it->second;
+
+    // coordination
+    vassert(
+      delta.type == topic_table_delta_type::replicas_updated
+        || delta.type == topic_table_delta_type::disabled_flag_updated,
+      "topic {} altered with topic_table_delta_type={} during "
+      "migration {}",
+      nt,
+      delta.type,
+      migration_id);
+    auto& mrstate = _mrstates.find(migration_id)->second;
+    auto& tstate = mrstate.outstanding_topics[nt];
+    clear_tstate_belongings(nt, tstate);
+    tstate.outstanding_partitions.clear();
+    // We potentially re-enqueue an already coordinated partition here.
+    // The first RPC reply will clear it.
+    co_await reconcile_topic(
+      nt, tstate, migration_id, mrstate.sought_state, false);
+
+    // local work
+    auto maybe_assignment = _topic_table.get_partition_assignment(delta.ntp);
+    bool has_local_replica = false;
+    if (maybe_assignment) {
+        for (const auto& replica : maybe_assignment->replicas) {
+            if (_self == replica.node_id) {
+                has_local_replica = true;
+                break;
+            }
+        }
```

Comment:
this can be replaced with utility function:  `inline bool contains_node(
  const std::vector<model::broker_shard>& replicas, model::node_id id)`